sdk/communication/AzureCommunicationChat/Source/ChatClient.swift

```diff
@@ -330,4 +333,154 @@ public class ChatClient {
+        let transport = RegistrarTransportSettings(
+            ttl: RegistrarSettings.ttl,
+            path: deviceToken,
+            context: RegistrarSettings.context
+        )
+
+        // Register for push notifications
+        registrarClient.setRegistration(with: clientDescription, for: [transport]) { result in
+            switch result {
+            case let .success(response):
+                self.pushNotificationsStarted = true
+                completionHandler(.success(response))
+            case let .failure(error):
+                self.options.logger
+                    .error("Failed to start push notifications with error: \(error.localizedDescription)")
+                completionHandler(.failure(AzureError.client("Failed to start push notifications", error)))
+            }
+        }
+    }
+
+    /// Stop push notifications.
+    /// - Parameter completionHandler: Success indicates push notifications have been stopped.
+    public func stopPushNotifications(
+        completionHandler: @escaping (Result<HTTPResponse?, AzureError>) -> Void
+    ) {
+        guard pushNotificationsStarted else {
+            completionHandler(.failure(AzureError.client("Push notifications are not enabled.")))
+            return
+        }
+
+        // Recreate the registrarClient if it doesn't exist
+        if registrarClient == nil {
+            // Internal options do not use the CommunicationSignalingErrorHandler
+            let internalOptions = AzureCommunicationChatClientOptionsInternal(
+                apiVersion: AzureCommunicationChatClientOptionsInternal.ApiVersion(options.apiVersion),
+                logger: options.logger,
+                telemetryOptions: options.telemetryOptions,
+                transportOptions: options.transportOptions,
+                dispatchQueue: options.dispatchQueue
+            )
+
+            let communicationCredential = TokenCredentialAdapter(credential)
+            let authPolicy = BearerTokenCredentialPolicy(credential: communicationCredential, scopes: [])
+
+            // Initialize the RegistrarClient
+            guard let tempRegistrarClient = try? RegistrarClient(
+                endpoint: RegistrarSettings.endpoint,
+                credential: credential,
+                registrationId: registrationId,
+                authPolicy: authPolicy,
+                withOptions: internalOptions
+            ) else {
+                completionHandler(.failure(
+                    AzureError
+                        .client("Failed to stop push notifications. Fail to find or create registrarClient.")
+                ))
+                return
+            }
+
+            registrarClient = tempRegistrarClient
+        }
+
+        registrarClient!.deleteRegistration { result in
+            switch result {
+            case let .success(response):
+                self.pushNotificationsStarted = false
+                completionHandler(.success(response))
```

Comment:
You also need to generate a new registrationId for a subsequent call to startPushNotifications. Otherwise it will keep trying to use the same one, which I would think is not what you want. If it is, your previous code wouldn't have done that.

Reply:
Thanks. Fixed and will push the change later.